perf/rdm_latency.c

```diff
@@ -0,0 +1,429 @@
+int skip_large = 10;
+int loop_large = 100;
+int large_message_size = 8192;
+
+char s_buf_original[MYBUFSIZE];
+char r_buf_original[MYBUFSIZE];
+
+static int rx_depth = 512;
+
+struct fi_info *fi, *hints;
+struct fid_fabric *fab;
+struct fid_domain *dom;
+struct fid_ep *ep;
+struct fid_av *av;
+struct fid_cq *rcq, *scq;
+struct fid_mr *mr;
+struct fi_context fi_ctx_send;
+struct fi_context fi_ctx_recv;
+struct fi_context fi_ctx_av;
+
+void *addrs;
+fi_addr_t *fi_addrs;
+
+typedef struct buf_desc {
+	uint64_t addr;
+	uint64_t key;
+} buf_desc_t;
+
+buf_desc_t *rbuf_descs;
+
+int myid, numprocs;
+
+void print_usage(void)
+{
+	fprintf(stderr, "\nOptions:\n");
+	fprintf(stderr, "  -f <provider>\tspecific provider name eg sockets, verbs\n");
+	fprintf(stderr, "  -h\t\tdisplay this help output\n");
+	return;
+}
+
+static double get_time(void)
+{
+	struct timeval tv;
+	double td;
+	gettimeofday(&tv, NULL);
+	td = tv.tv_sec + (tv.tv_usec * 0.000001);
+	return td;
+}
+
+static void free_ep_res(void)
+{
+	fi_close(&av->fid);
+	fi_close(&rcq->fid);
+	fi_close(&scq->fid);
+}
+
+static int alloc_ep_res(void)
+{
+	struct fi_cq_attr cq_attr;
+	struct fi_av_attr av_attr;
+	int ret;
+
+	memset(&cq_attr, 0, sizeof cq_attr);
+	cq_attr.format = FI_CQ_FORMAT_CONTEXT;
+	cq_attr.wait_obj = FI_WAIT_NONE;
+	cq_attr.size = rx_depth;
+
+	/* Open completion queue for send completions */
+	ret = fi_cq_open(dom, &cq_attr, &scq, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_cq_open", ret);
+		goto err1;
+	}
+
+	/* Open completion queue for recv completions */
+	ret = fi_cq_open(dom, &cq_attr, &rcq, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_cq_open", ret);
+		goto err2;
+	}
+
+	memset(&av_attr, 0, sizeof av_attr);
+	av_attr.type = fi->domain_attr->av_type ?
+			fi->domain_attr->av_type : FI_AV_TABLE;
+	av_attr.count = 2;
+	av_attr.name = NULL;
+
+	/* Open address vector (AV) for mapping address */
+	ret = fi_av_open(dom, &av_attr, &av, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_av_open", ret);
+		 goto err3;
+	 }
+
+	return 0;
+
+err3:
+	fi_close(&rcq->fid);
+err2:
+	fi_close(&scq->fid);
+err1:
+	return ret;
+}
+
+static int bind_ep_res(void)
+{
+	int ret;
+
+	/* Bind Send CQ with endpoint to collect send completions */
+	ret = fi_ep_bind(ep, &scq->fid, FI_SEND);
+	if (ret) {
+		FT_PRINTERR("fi_ep_bind", ret);
+		return ret;
+	}
+
+	/* Bind Recv CQ with endpoint to collect recv completions */
+	ret = fi_ep_bind(ep, &rcq->fid, FI_RECV);
+	if (ret) {
+		FT_PRINTERR("fi_ep_bind", ret);
+		return ret;
+	}
+	
+	/* Bind AV with the endpoint to map addresses */
+	ret = fi_ep_bind(ep, &av->fid, 0);
+	if (ret) {
+		FT_PRINTERR("fi_ep_bind", ret);
+		return ret;
+	}
+
+	ret = fi_enable(ep);
+	if (ret) {
+		FT_PRINTERR("fi_enable", ret);
+		return ret;
+	 }
+
+	return ret;
+}
+
+static int init_fabric(void)
+{
+	int ret;
+	uint64_t flags = 0;
+
+	/* Get fabric info */
+	ret = fi_getinfo(FT_FIVERSION, NULL, NULL, flags, hints, &fi);
+	if (ret) {
+		FT_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
```

Comment:
Can we add options for selecting node name(/port)? Just like -s option in fabtests.
This will help to run these tests on nodes with multiple interfaces.